Mathlib/Data/Nat/Cast/Defs.lean

```diff
@@ -0,0 +1,247 @@
+/-
+Copyright (c) 2014 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Gabriel Ebner
+-/
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Algebra.NeZero
+import Mathlib.Tactic.SplitIfs
+
+/-!
+# Cast of natural numbers
+
+This file defines the *canonical* homomorphism from the natural numbers into an
+`AddMonoid` with a one.  In additive monoids with one, there exists a unique
+such homomorphism and we store it in the `natCast : ℕ → R` field.
+
+Preferentially, the homomorphism is written as the coercion `Nat.cast`.
+
+## Main declarations
+
+* `NatCast`: Type class for `Nat.cast`.
+* `AddMonoidWithOne`: Type class for which `Nat.cast` is a canonical monoid homomorphism from `ℕ`.
+* `Nat.cast`: Canonical homomorphism `ℕ → R`.
+-/
+
+
+universe u
+
+/-- The numeral `((0+1)+⋯)+1`. -/
+protected def Nat.unaryCast {R : Type u} [One R] [Zero R] [Add R] : ℕ → R
+  | 0 => 0
+  | n + 1 => Nat.unaryCast n + 1
+#align nat.unary_cast Nat.unaryCast
+
+/-- Type class for the canonical homomorphism `ℕ → R`. -/
+class NatCast (R : Type u) where
+  /-- The canonical map `ℕ → R`. -/
+  protected natCast : ℕ → R
+#align has_nat_cast NatCast
+#align has_nat_cast.nat_cast NatCast.natCast
+
+/-- Canonical homomorphism from `ℕ` to a additive monoid `R` with a `1`.
+This is just the bare function in order to aid in creating instances of `AddMonoidWithOne`. -/
+@[coe]
+protected def Nat.cast {R : Type u} [NatCast R] : ℕ → R :=
+  NatCast.natCast
+#align nat.cast Nat.cast
+
+-- see note [coercion into rings]
+instance [NatCast R] : CoeTail ℕ R where coe := Nat.cast
+
+-- see note [coercion into rings]
+instance [NatCast R] : CoeHTCT ℕ R where coe := Nat.cast
+
+-- I *think* we want this instance? Otherwise we can't write `2 : R` (see `Nat.cast_two` below)
+-- this was not in the Lean 3 code
+instance [NatCast R] (n : ℕ) : OfNat R n where ofNat := Nat.cast n
```

Comment:
Perhaps provide a doc-string on this instance foreshadowing the simp lemma `Nat.cast_ofNat` below, to explain what the simp norm form is.

Reply:
Thanks, you made me realize that the extra bits at the bottom should have been moved up, and they included this `OfNat` instance.

I *think* if I understood Mario correctly yesterday, that `Nat.cast` is the `simp` normal form, but for literals you always have to have `OfNat.ofNat`, so I don't think that `Nat.cast_ofNat` is dictating the `simp` normal form, but maybe I have terribly misunderstood something.